keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+objects (like storage classes).
+
+The downsides are:
+- Some attributes (driver name, topology) must be stored multiple times
+  compared to a more complex object, so overall data size in etcd is higher.
+- Higher number of objects which all need to be retrieved by a client
+  which does not already know which `CSIStorageCapacity` object it is
+  interested in.
+
+##### Example: local storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-ab96d356-0d31-11ea-ade1-8b7e883d1af1
+spec:
+  driverName: hostpath.csi.k8s.io
+  storageClassName: some-storage-class
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values:
+        - node-1
+status:
+  capacity: 256G
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-c3723f32-0d32-11ea-a14f-fbaf155dff50
+spec:
+  driverName: hostpath.csi.k8s.io
+  storageClassName: some-storage-class
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values:
+        - node-2
+status:
+  capacity: 512G
+```
+
+##### Example: affect of storage classes
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-9c17f6fc-6ada-488f-9d44-c5d63ecdf7a9
+spec:
+  driverName: lvm
+  storageClassName: striped
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values:
+        - node-1
+status:
+  capacity: 256G
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-f0e03868-954d-11ea-9d78-9f197c0aea6f
+spec:
+  driverName: lvm
+  storageClassName: mirrored
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: kubernetes.io/hostname
+        operator: In
+        values:
+        - node-1
+status:
+  capacity: 128G
+```
+
+##### Example: network attached storage
+
+```
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-b0963bb5-37cf-415d-9fb1-667499172320
+spec:
+  driverName: pd.csi.storage.gke.io
+  storageClassName: some-storage-class
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: topology.kubernetes.io/region
+        operator: In
+        values:
+        - us-east-1
+status:
+  capacity: 128G
+
+apiVersion: storage.k8s.io/v1alpha1
+kind: CSIStorageCapacity
+metadata:
+  name: csisc-64103396-0d32-11ea-945c-e3ede5f0f3ae
+spec:
+  driverName: pd.csi.storage.gke.io
+  storageClassName: some-storage-class
+  nodeTopology:
+    nodeSelectorTerms:
+    - matchExpressions:
+      - key: topology.kubernetes.io/region
+        operator: In
+        values:
+        - us-west-1
+status:
+  capacity: 256G
+```
+
+#### CSIDriver.spec.storageCapacity
+
+A new field `storageCapacity` of type `boolean` with default `false`
```

Comment:
Do we need this new field? Or would the presence of the CSIStorageCapacity object be sufficient?

 I guess without the field, then rollout would potentially not be as smooth? With the field, you could upgrade the DaemonSet to the new version and only flip the driver bit when there is capacity reported across all nodes.

Reply:
> Do we need this new field? Or would the presence of the CSIStorageCapacity object be sufficient?

Suppose there is not a single `CSIStorageCapacity` for the driver: is that because the driver does not support the feature, or because capacity information has not been created yet (driver still starting up)? The scheduler cannot distinguish between the two cases.

If it assumes that "no objects means no capacity", then it'll stop scheduling for drivers which don't support the feature.

If it assumes that "no objects means feature not supported", then it'll schedule instead of waiting for information. That will work eventually (it's just another variant of "acting on incomplete or out-dated information" and the recovery mechanisms need to deal with it), though, so perhaps that flag can indeed be dropped?

> I guess without the field, then rollout would potentially not be as smooth? With the field, you could upgrade the DaemonSet to the new version and only flip the driver bit when there is capacity reported across all nodes.

I was thinking of a simpler deployment where the driver just gets installed once through a set of yamls and then Kubernetes itself waits for the driver to provide some capacity information, without an admin or operator implementing such an additional post-install step.

Reply:
I now think that we don't need the new field, without any loss of functionality. Here's what we can do instead:
- Support the manual creation of CSIStorageCapacity objects by ensuring that external-provisioner only deletes objects that it also created. This was implied by the design, but not called out explicitly yet. I expect that this will be useful also for the autoscaler because then a cluster admin can use such manually created objects to define (local) storage that currently non-existent nodes will have access to.
- The scheduler blocks pod scheduling if it finds at least one CSIStorageCapacity object for the CSI driver, but none that has enough capacity.
- CSI driver deployments create one CSIStorageCapacity object, for example with the CSI driver name as name, if the deployment is going to produce storage capacity information. That object can define zero available storage, topology doesn't matter. All that matters is that the object exists and has the CSI driver name.

@msau42: what do you think?

Reply:
This may make upgrading a local storage driver in a large cluster difficult, say 5000 nodes. While the rolling update is happening, the scheduler will go from 5000 node choices quickly down to a handful. This could cause a lot of scheduling failures while this upgrade is happening. Ideally, we would want some way to switch over the behavior after the upgrade is complete.

Reply:
@msau42: This is about upgrading from a driver deployment which has no support for storage capacity to one which does, right? In that case, storage capacity constraints will kick in once the first `CSIStorageCapacity` object appears, and you are worried that it would be better to let the driver continue with creating such objects for a while.

Let me augment my proposal to cover this: to opt into storage capacity scheduling, a driver deployment *must* create a `CSIStorageCapacity` with the driver name as object name. Whoever deploys the driver can then decide to create that object immediately or wait a bit.

Reply:
We discussed this today and agreed that the `CSIDriverSpec.StorageCapacity` is the better API, so we'll keep the KEP as originally proposed and merged.